Adding functions for agnostic tensor manipulations

josh146 opened the pull request:
**Context:**

To support templates in tape-mode, all classical processing within the QNode must be done in a framework agnostic manner. This motivates a unified API for manipulating NumPy, Autograd, Torch, and TensorFlow tensors.

**Description of changes:**

* Adds the `TensorBox` class, a *container* for array-like objects that allows array manipulation to be performed in a unified manner for supported linear algebra packages.

  The `TensorBox` class is designed for constructing a unified tensor manipulation API.

* The `TensorBox` class is an abstract class, that defines various methods, such as `.abs()`, `.stack()`, etc. By containing an array-like object within a `TensorBox`, method chaining can be used to perform manipulations:

  ```pycon
  >>> y.T.abs().transpose([1, 0, 2]).sin()
  ```

  By wrapping array-like objects in a ``TensorBox`` class, array manipulations are performed by simply chaining method calls. Under the hood, the method call is dispatched to the corresponding linear algebra library based on the wrapped array type, without the need to import any external libraries manually. As a result, autodifferentiation is preserved where needed.

* Various subclasses implement the abstract methods for the different interfaces; `NumpyBox`, `AutogradBox`, `TorchBox`, and `TensorFlowBox`.

* Finally, a collection of functions are provided to wrap the new `TensorBox` functionality, avoiding the need to manually work with `TensorBox` objects at all. These functions should be used in templates, collections, and optimizers, to ensure autodifferentiation is preserved.

  For example, the `qml.tensorbox.stack()` function works identically regardless the input type of the tensors, by dispatching to the correct framework.

  Torch:

  ```pycon
  >>> x = torch.tensor([1., 2.])
  >>> qml.tensorbox.ones_like(x)
  tensor([1, 1])
  ```

  TensorFlow:

  ```pycon
  >>> y = tf.Variable([[0], [5]])
  >>> qml.tensorbox.ones_like(y, dtype=np.complex128)
  <tf.Tensor: shape=(2, 1), dtype=complex128, numpy=
  array([[1.+0.j],
         [1.+0.j]])>
  ```

  For more details see the documentation: https://pennylane--886.org.readthedocs.build/en/886/code/qml_tensorbox.html

**Benefits:**

* Provides convenience functions for manipulating tensors in a framework agnostic manner, while preserving autodifferentiation.

**Drawbacks:**

* Only a very limited set of abstract methods are currently implemented. These are the *bare minimum* required to allow the templates to work in tape mode; more methods can be added on an as-needed basis.

mariaschuld commented on pennylane/tensorbox/tensorbox.py:
```diff
@@ -0,0 +1,253 @@
+        if isinstance(tensor, (list, tuple)) or namespace == "numpy":
+            from .numpy_box import NumpyBox
+
+            return NumpyBox.__new__(NumpyBox, tensor)
```
So if it is a "normal" numpy array we use autograd, but if it is a list or tuple we don't because we assume it is not diffable anyways? 

A bit tangential here but I started wondering how clear this is to new users...

josh146 replied:
Not quite:

* If the input is a list, tuple, or `np.ndarray`, we create a `NumpyBox`, which uses vanilla numpy for all manipulations.

* If the input is a PennyLane tensor, we create an `AutogradBox`.

The reason being that autograd (a) has a bit of overhead compared to standard numpy, so no point using it when we don't have to, and (b) autograd is a bit restricted - not all numpy functions are supported, so sometimes you must do the manipulation in python rather than numpy (which adds more slow down)

mariaschuld commented on pennylane/tensorbox/tensorbox.py:
```diff
@@ -0,0 +1,253 @@
+        self._data = tensor
+
+    def __repr__(self):
+        return f"<TensorBox {self.unbox().__repr__()}>"
```
Nice

mariaschuld commented on pennylane/tensorbox/tensorbox.py:
```diff
@@ -0,0 +1,253 @@
+    @property
+    @abc.abstractmethod
+    def interface(self):
+        """str, None: The linear algebra package that the :class:`.TensorBox` class
```
Maybe better just to say "package" since it is not all linear algebra they do?

josh146 replied:
```suggestion
        """str, None: The package that the :class:`.TensorBox` class
```

mariaschuld commented on pennylane/tensorbox/numpy_box.py:
```diff
@@ -0,0 +1,65 @@
+
+        super().__init__(tensor)
+
+    def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
```
Curious what this is...maybe add a docstring since it does not seem to be defined in a super class?

josh146 replied:
done!

co9olguy replied:
I didn't spot the new docstring. Can you point out where it was added?

josh146 replied:
I hadn't pushed it 🙈 I've now pushed it!

mariaschuld commented on pennylane/tensorbox/__init__.py:
```diff
@@ -0,0 +1,19 @@
+"""
+This package contains the ``TensorBox`` class, a container for array-like objects
+that allows array manipulation to be performed in a unified manner for supported
+linear algebra packages.
```
Shall we mention that it is loosely inspired by eagerpy?

co9olguy replied:
I found that this docstring actually doesn't make it clear enough what it is doing. Could we mention that this class handles the dispatching of function calls to the relevant libraries?

josh146 replied:
I've re-written this docstring, and replaced all 'linear algebra packages' with 'tensor/array manipulation frameworks' :)

mariaschuld commented on tests/tensorbox/test_tensorbox.py:
```diff
@@ -0,0 +1,541 @@
+        """Test that an exception is raised if the input type
+        is unknown"""
+        with pytest.raises(ValueError, match="Unknown tensor type"):
+            qml.TensorBox(True)
```
Nice!

mariaschuld commented on tests/tensorbox/test_tensorbox.py:
```diff
@@ -0,0 +1,541 @@
+        """Test length"""
+        x = np.array([[1, 2], [3, 4]])
+        res = qml.TensorBox(x)
+        assert len(res) == len(x) == 2
```
That is so useful!

mariaschuld commented on tests/tensorbox/test_tensorbox.py:
```diff
@@ -0,0 +1,541 @@
+    def test_ufunc_compatibility(self):
+        """Test that the NumpyBox class has ufunc compatibility"""
+        x = np.array([0.1, 0.2, 0.3])
+        res = np.sum(np.sin(qml.TensorBox(x)))
```
Oh ufunc is for vectorised evalaution?

josh146 replied:
Yep, `ufunc` is a NumPy-aware function that automatically broadcasts :)

mariaschuld commented on tests/tensorbox/test_tensorbox.py:
```diff
@@ -0,0 +1,541 @@
+
+        res = tape.gradient(y, x)
+        expected = tf.constant([[0.0, 0.0, 0.0], [8.0, 0.0, 0.0]])
+        assert np.allclose(res, expected)
```
I don't want to be pedantic, but the biggest danger in this class is that the different interfaces implement the same function differently. I don't remember a good example, but had it before that how dims/axes and other args are interpreted is not 100% the same...well, didn't we have that problem with tolerances before?

So, I wonder if it is worth adding a test class that checks whether the numpy results are all the same, using some edge test cases?

mariaschuld replied:
Do we need a test for chaining?

josh146 replied:
@mariaschuld I've now added a new test suite (`test_functions.py`) that tests the functions using parametrized test data from all interfaces, to make sure they agree on the result!

mariaschuld reviewed (commented):
This is so nice Josh!

The only thing that would be nice is to add this module to the docs and make a changelog entry? Will also be useful to check the docstrings before final approval.

josh146 commented:
Thanks Maria for your speedy comments! I'll address them in the morning.

Note that the documentation is visible here: https://pennylane--886.org.readthedocs.build/en/886/code/api/pennylane.TensorBox.html

Since it's imported top level, the module didn't need to be added separately

mariaschuld reviewed (approved):
Thanks for the link, checked docs and they look fine!

(Maybe nit picking: the `numpy` function looks strange with the empty return. Maybe one could just add a short explanation?

```
Return:

Return type:
   ndarray
```

)

trbromley commented:
[ch1037]

co9olguy commented:
Would we eventually expose these functions to the user in a UI that looks more natural? e.g., `qml.sin(t)` instead of `t.sin()`?

josh146 commented:
> Would we eventually expose these functions to the user in a UI that looks more natural? e.g., qml.sin(t) instead of t.sin()?

These will not be exposed to the user, as we decided in the classical processing ADR. They will only be used internally by developers to write templates/cost functions that are interface agnostic

co9olguy commented:
I think that's a feature that (power) users will want to access. Any barriers to implementing a nicer UI, even for developers?

co9olguy commented on pennylane/tensorbox/tensorbox.py:
```diff
@@ -0,0 +1,253 @@
+        >>> x = tf.Variable([0.4, 0.1, 0.5])
+        >>> a = tf.constant([1., 2., 3.])
+        >>> y = TensorBox(x)
+        >>> y.stack([a, y])
```
:no_good:

josh146 replied:
See `qml.tensorbox.stack([a, y])` 🙂

co9olguy commented on pennylane/tensorbox/autograd_box.py:
```diff
@@ -0,0 +1,56 @@
+
+    @property
+    def T(self):
+        return AutogradBox(self.unbox().T)
```
There's a lot of unboxing and reboxing here. I wonder if there's a way to do it in-place? :thinking:

co9olguy commented on pennylane/tensorbox/numpy_box.py:
```diff
@@ -0,0 +1,65 @@
+
+    @property
+    def T(self):
+        return NumpyBox(self.unbox().T)
```
Other than the name of the "Box", there is a lot of very similar code for some of these functions between the classes. I wonder if there is a cleaner way to do it, e.g., by avoiding unboxing/reboxing and replacing these functions to just point to the appropriate interface's function directly (when appplicable)? :thinking:

co9olguy commented on pennylane/tensorbox/tf_box.py:
```diff
@@ -0,0 +1,57 @@
+    def interface(self):
+        return "tf"
+
+    def __len__(self):
```
How come this class had to overwrite that dunder method?

mariaschuld replied:
I don't think `len` is defined for tensorflow objects...was struggling with that before. 

As a good rule of thumb: wrapping has lots of nasty surprised :)

josh146 replied:
It's worse than that --- it is defined for `tf.Tensor` but not `tf.Variable` or `tf.ResourceVariable`! So it works sometimes, but not all the time, which is really annoying for wrapping.

co9olguy commented on tests/tensorbox/test_tensorbox.py:
```diff
@@ -0,0 +1,541 @@
+        np = qml.numpy
+        x = np.array([[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]])
+
+        cost_fn = lambda a: (qml.TensorBox(a).T ** 2).unbox()[0, 1]
```
Will a user always have to unbox if they want to do something like this?

josh146 replied:
With the new functional approach, the unboxing is no longer visible

co9olguy commented on tests/tensorbox/test_tensorbox.py:
```diff
@@ -0,0 +1,541 @@
+
+    def test_expand_dims(self, autograd):
+        """Test that dimension expansion works"""
+        from pennylane.tensorbox.autograd_box import AutogradBox
```
Could we just import these at the class level, rather than in every test?

josh146 replied:
Unfortunately not; python executes all class attributes at module import, which would then fail if the corresponding package was not installed

josh146 replied:
I've split this into separate files, one for each class. This way we can do the imports at the top.

co9olguy reviewed (changes requested):
Thanks @josh146! Solid work.

My biggest question is about the interface. Even for something developer-facing, it feels overly clunky. Specific examples are things like
```res = y.unwrap([y, z])```
or the need to explicitly convert everything to a `TensorBox` in order to use this functionality (would be nicer if things were automatically determined by functional calls on interface-dependent tensors/arrays)

co9olguy commented:
> By utilizing method chaining, we avoid issues with function dispatch (something like fn.ones([4, 2]) would recieve no array-like input to help with dispatching, but a static method qml.TensorBox(array).ones([4, 2]) works nicely)
This benefit wasn't too clear to me. If something is showing up as a bare list, why do we need to dispatch. Wouldn't standard numpy be sufficient as a default? (since all frameworks accept numpy inputs, and it is unlikely that a user would want to make a list differentiable without explicitly declaring it so)

josh146 commented on pennylane/tensorbox/tensorbox.py:
```diff
@@ -0,0 +1,253 @@
+    >>> y.ones_like()
+    tf.Tensor([1. 1. 1.], shape=(3,), dtype=float32)
+
+    Unless specified, the returned tensors are also unified tensors, allowing
```
```suggestion
    Unless specified, the returned tensors are also ``TensorBox`` instances, allowing
```

josh146 commented on pennylane/tensorbox/tensorbox.py:
```diff
@@ -0,0 +1,253 @@
+    for method chaining:
+
+    >>> y.ones_like().expand_dims(0)
+    tf.Tensor([[1. 1. 1.]], shape=(1, 3), dtype=float32)
```
```suggestion
    tf.Tensor([[1., 1., 1.]], shape=(1, 3), dtype=float32)
```

josh146 commented on pennylane/tensorbox/tensorbox.py:
```diff
@@ -0,0 +1,253 @@
+        >>> y = TensorBox(x)
+        >>> z = tf.constant([0.1, 0.2])
+
+        Note that this is a static method, so we must pass the unified tensors itself
```
```suggestion
        Note that this is a static method, so we must pass the tensor represented by the ``TensorBox``
```

josh146 commented on pennylane/tensorbox/tensorbox.py:
```diff
@@ -0,0 +1,253 @@
+        Note that this is a static method, so we must pass the unified tensors itself
+        if we would like it to be included.
+
+        >>> res = y.unwrap([y, z])
```
@co9olguy I've now included a file with front-facing, functional calls that look and feel a lot better. So these tensorbox methods are only ever used in the background.

josh146 commented on pennylane/tensorbox/tensorbox.py:
```diff
@@ -0,0 +1,253 @@
+    @abc.abstractmethod
+    def numpy(self):
+        """Converts the tensor to a standard, non-differentiable NumPy ndarray or Python scalar if
+        the tensor is 0-dimensional.
```
TF and Torch appear to have the following behaviour:

```pycon
>>> x = tf.constant(0.5)
>>> x.ndim
0
>>> x.numpy()
0.5
```

That is, 0-dimensional tensors will get converted into scalars, not `array(0.5)`.

josh146 replied:
So I've replicated this behaviour here.

co9olguy replied:
I was more asking about if the tensor didn't have shape `()`. The way this docstring was worded made it sound like all tensors would be zero-dimensional (i.e., scalars)

josh146 commented on pennylane/tensorbox/tensorbox.py:
```diff
@@ -0,0 +1,253 @@
+        the tensor is 0-dimensional.
+
+        Returns:
+            array:
```
```suggestion
            array, float, int: NumPy ndarray, or Python scalar if the input is 0-dimensional
```

josh146 replied:
> looks weird to have the colon but nothing after 🤔

(I noticed the other day that this is quite common in other packages if the return is 'obvious' from the docstring summary)

co9olguy replied:
weird, would make more sense if the colon wasn't there

josh146 commented on pennylane/tensorbox/tensorbox.py:
```diff
@@ -0,0 +1,253 @@
+        >>> y = TensorBox(x)
+        >>> y.ones_like()
+        tf.Tensor(
+        [[1. 1.]
+         [1. 1.]], shape=(2, 2), dtype=float32)
```
I copied and pasted these from my terminal, so it looks like TF just doesn't print commas 🤔

josh146 commented on pennylane/tensorbox/tensorbox.py:
```diff
@@ -0,0 +1,253 @@
+        return f"<TensorBox {self.unbox().__repr__()}>"
+
+    def __len__(self):
+        return len(self.unbox())
```
Good idea - I've removed the `unbox()` method calls everywhere, and replaced them with the `data` attribute.

josh146 commented on pennylane/tensorbox/tensorbox.py:
```diff
@@ -0,0 +1,253 @@
+
+    @staticmethod
+    def unbox_list(tensors):
+        """Unboxes or unwraps a list of array like objects, converting any :class:`TensorBox`
```
I've standardized this everywhere, thanks for catching! I'm using the same standard as the numpy docs. They use `array_like` as the arg/returns type, and `array-like` in text.

josh146 commented:
@co9olguy @mariaschuld thanks for all the suggestions! I've implemented them, the PR is now ready for another review.

In particular, @co9olguy I've implemented your idea for a functional UX, it ends up working quite nicely:

```pycon
>>> x = torch.tensor([1., 2.])
>>> qml.tensorbox.ones_like(x)
tensor([1, 1])
```

TensorFlow:

```pycon
>>> y = tf.Variable([[0], [5]])
>>> qml.tensorbox.ones_like(y, dtype=np.complex128)
<tf.Tensor: shape=(2, 1), dtype=complex128, numpy=
array([[1.+0.j],
       [1.+0.j]])>
```

co9olguy reviewed (approved):
Very thorough work, awesome @josh146!

mariaschuld commented on pennylane/tensorbox/__init__.py:
```diff
@@ -0,0 +1,52 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
```
just  a thought, while the module name 'tensorbox' is logical, there may be a better one describing what it does? such as   `processing` or `maths` or `tensor`?

josh146 replied:
I would rather not `maths` (as it sounds too useful/inviting for a user, who might explore `qml.maths`), or `tensor` (as we already have so many things called `tensor` in the codebase).

How about `qml.proc`? Short and sweet (`qml.proc.stack([x, y])`), but less inviting 😆

mariaschuld replied:
what about `qml.processing`? if name is too long one can just import the module?

josh146 replied:
I think we should avoid importing this module directly, since it will likely be used in so many places we will quickly end up in circular dependency hell 😆 

So we should always do
```python
import pennylane as qml
...
qml.processing.stack([x, y])
```

co9olguy replied:
Came here to say this, thanks @mariaschuld

josh146 replied:
Done!

mariaschuld commented on pennylane/tensorbox/__init__.py:
```diff
@@ -12,8 +12,41 @@
+* NumPy
+* Autograd
+* TensorFlow
+* PyTorch
```
actualle, in the built docs they appear on first screen, and seems ok

mariaschuld commented on pennylane/tensorbox/autograd_box.py:
```diff
@@ -0,0 +1,61 @@
+        return "autograd"
+
+    def numpy(self):
+        return self.data.numpy()
```
wouldn't self.data be a numpy array here?

josh146 replied:
Nope, `self.data` is the raw PennyLane NumPy tensor. While still a subclass of `ndarray`, so in most cases duck typing works fine, there are a few rare places where you need an actual `ndarray` representation.

mariaschuld commented on pennylane/tensorbox/fn.py:
```diff
@@ -0,0 +1,416 @@
+    Returns:
+        .TensorBox: A TensorBox that will dispatch to the correct framework
+        given the rules of precedence. This TensorBox will contain the *first*
+        tensor like object in ``values`` that corresponds to the highest-priority
```
```suggestion
        tensor-like object in ``values`` that corresponds to the highest-priority
```

this statement is a bit confusing?

mariaschuld replied:
to be more precise: the grammar makes it unclear if values corresponds to the framework, or the object

mariaschuld commented on pennylane/tensorbox/fn.py:
```diff
@@ -0,0 +1,416 @@
+        framework.
+
+    To determine the framework to dispatch to, the following rules
+    are followed:
```
```suggestion
    are applied:
```

mariaschuld commented on pennylane/tensorbox/fn.py:
```diff
@@ -0,0 +1,416 @@
+    * Autograd tensors *may* be present alongside Torch and TensorFlow tensors,
+      but Torch and TensorFlow take precendence; the autograd arrays will
+      be treated as non-differentiable NumPy arrays. A warning will be raised
+      suggesting that vanilla NumPy be used instead.
```
good idea

mariaschuld commented on pennylane/tensorbox/fn.py:
```diff
@@ -0,0 +1,416 @@
+        return TensorBox(values[interfaces.index("torch")])
+
+    if "autograd" in interfaces:
+        return TensorBox(values[interfaces.index("autograd")])
```
so this function takes a seq of objects, and joins only the objects of the priority interface to a new tensorbox? the other ones get dropped? isn't that unexpected?

josh146 replied:
Good question.

The idea here was not so much to return a wrapped tensorbox for every value in the sequence; we just need a _single_ tensorbox so that we know what interface to use. Once we have the right tensorbox, it's just a matter of static method chaining.

So the value of the returned tensorbox isn't really important; it could even be empty. Perhaps worth having this function return an 'empty' tensorbox, as long as it dispatches to the correct framework?

mariaschuld commented on pennylane/tensorbox/fn.py:
```diff
@@ -0,0 +1,416 @@
+    return np.allclose(t1, t2, rtol=rtol, atol=atol, **kwargs)
+
+
+allclose.__doc__ = np.allclose.__doc__
```
clever!

mariaschuld commented on pennylane/tensorbox/fn.py:
```diff
@@ -0,0 +1,416 @@
+    >>> get_interface(x)
+    'autograd'
+    """
+    return TensorBox(tensor).interface
```
that feels like a good way to evaluate the interface!

mariaschuld commented on pennylane/tensorbox/fn.py:
```diff
@@ -0,0 +1,416 @@
+def toarray(tensor):
+    """Returns the tensor as a NumPy ``ndarray``. No copying
+    is performed; the tensor and the returned array share the
+    same storage.
```
are you referring to 'in-place' manipulation?

mariaschuld replied:
if the user calls this function without catching the returned object, will tensor be overwritten?

josh146 replied:
No not quite, more so that the underlying memory representing the array data is the same.

This is how numpy/torch/tf tend to work a lot under the hood (see https://numpy.org/doc/stable/reference/generated/numpy.ndarray.view.html)

mariaschuld commented on pennylane/tensorbox/fn.py:
```diff
@@ -0,0 +1,416 @@
+    False
+    >>> with tf.GradientTape() as tape:
+    ...     print(requires_grad(x))
+    True
```
just a thought...wouldn't it be important to support the same call structure in all frameworks? if i don't know which tensor type enters a template, how do i know whether to call `requires_grad()` or `.requires_grad`?

mariaschuld commented on pennylane/tensorbox/fn.py:
```diff
@@ -0,0 +1,416 @@
+    >>> shape(x)
+    (2, 3)
+    """
+    return TensorBox(tensor).shape
```
this will be SO useful

mariaschuld commented on pennylane/tensorbox/tensorbox.py:
```diff
@@ -0,0 +1,353 @@
+
+    @abc.abstractmethod
+    def numpy(self):
+        """Converts the tensor to a standard, non-differentiable NumPy ndarray or Python scalar if
+        the tensor is 0-dimensional.
```
```suggestion
        """Converts the tensor to a standard, non-differentiable NumPy ndarray, or to a Python scalar if
        the tensor is 0-dimensional.
```maybe this makes it clearer?

mariaschuld commented on tests/tensorbox/test_autograd_box.py:
```diff
@@ -0,0 +1,194 @@
+
+    res = qml.tensorbox.TensorBox(x).cast(np.float64)
+    expected = np.array([1.0, 2.0, 3.0])
+    assert np.all(res == expected)
```
interesting, i did not know `==` picks up the type

mariaschuld replied:
wouldn't it be safer to check the dtype of the elements in `res`?

josh146 replied:
Added!

mariaschuld commented on tests/tensorbox/test_autograd_box.py:
```diff
@@ -0,0 +1,194 @@
+    """Test length"""
+    x = np.array([[1, 2], [3, 4]])
+    res = qml.tensorbox.TensorBox(x)
+    assert len(res) == len(x) == 2
```
maybe test edge case of length 0?

josh146 replied:
Done!

mariaschuld commented on tests/tensorbox/test_autograd_box.py:
```diff
@@ -0,0 +1,194 @@
+    """Test that the AutogradBox class has ufunc compatibility"""
+    x = np.array([0.1, 0.2, 0.3])
+    res = np.sum(np.sin(qml.tensorbox.TensorBox(x)))
+    assert res.unbox().item() == np.sin(0.1) + np.sin(0.2) + np.sin(0.3)
```
don't want to be a pain, but since there are so many if statements in the ufunc catering for output types, wouldn't it be prudent to check more test cases here?

josh146 replied:
Done :)

mariaschuld commented on tests/tensorbox/test_functions.py:
```diff
@@ -0,0 +1,519 @@
+            fn._get_multi_tensorbox([x, y])
+
+    def test_return_tensorflow_box(self):
+        """Test that TensorFlow is correctly identified as the dispatching library"""
```
```suggestion
        """Test that TensorFlow is correctly identified as the dispatching library."""
```

mariaschuld commented on tests/tensorbox/test_functions.py:
```diff
@@ -0,0 +1,519 @@
+        assert res.interface == "tf"
+
+    def test_return_torch_box(self):
+        """Test that Torch is correctly identified as the dispatching library"""
```
```suggestion
        """Test that Torch is correctly identified as the dispatching library."""
```

mariaschuld commented on tests/tensorbox/test_functions.py:
```diff
@@ -0,0 +1,519 @@
+        assert res.interface == "torch"
+
+    def test_return_autograd_box(self):
+        """Test that autograd is correctly identified as the dispatching library"""
```
```suggestion
        """Test that autograd is correctly identified as the dispatching library."""
```

mariaschuld commented on tests/tensorbox/test_functions.py:
```diff
@@ -0,0 +1,519 @@
+        assert res.interface == "autograd"
+
+    def test_return_numpy_box(self):
+        """Test that NumPy is correctly identified as the dispatching library"""
```
```suggestion
        """Test that NumPy is correctly identified as the dispatching library."""
```

mariaschuld commented on tests/tensorbox/test_functions.py:
```diff
@@ -0,0 +1,519 @@
+
+@pytest.mark.parametrize("t1,t2", list(itertools.combinations(test_data, r=2)))
+def test_allequal(t1, t2):
+    """Test that the allequal function works for a variety
+    of input"""
```
```suggestion
    of inputs."""
```

mariaschuld replied:
```suggestion
    """Test that the allequal function works for a variety of inputs."""
```

mariaschuld commented on tests/tensorbox/test_functions.py:
```diff
@@ -0,0 +1,519 @@
+
+@pytest.mark.parametrize("t1,t2", list(itertools.combinations(test_data, r=2)))
+def test_allclose(t1, t2):
+    """Test that the allclose function works for a variety
+    of input"""
```
```suggestion
    """Test that the allclose function works for a variety of inputs."""
```

mariaschuld commented on tests/tensorbox/test_functions.py:
```diff
@@ -0,0 +1,519 @@
+    @pytest.mark.parametrize("t", test_data)
+    def test_cast_numpy(self, t):
+        """Test that specifying a NumPy dtype results in proper casting
+        behaviour"""
```
feels better if all docstrings finish with a period sign...?

josh146 replied:
😆 I'm always a bit more informal with test docstrings

mariaschuld commented on tests/tensorbox/test_functions.py:
```diff
@@ -0,0 +1,519 @@
+    """Tests for the expand_dims function"""
+
+    def test_expand_dims_sequence(self, shape, axis, new_shape):
+        """Test that the expand dimensions works correctly
+        when given a sequence"""
```
```suggestion
        """Test that expand_dimensions works correctly
        when given a sequence"""
```
here and below?

mariaschuld commented on tests/tensorbox/test_functions.py:
```diff
@@ -0,0 +1,519 @@
+
+
+interface_test_data = [
+    [(1, 2, 3), "numpy"],
```
what happens if we feed scalars? this might happen a lot actually!

josh146 replied:
Ah, feeding in scalars will raise an error currently, since the `TensorBox` won't be able to deduce which framework to dispatch to 🤔

mariaschuld commented on pennylane/tensorbox/fn.py:
```diff
@@ -0,0 +1,416 @@
+    is currently being watched by a gradient tape:
+
+    >>> x = tf.Variable([0.6, 0.1])
+    >>> requires_grad(x)
```
use `qml.tensorbox.requires_grad` as well?

josh146 replied:
Ah, I think I'll remove the `qml.tensorbox` from the ones above!

mariaschuld commented on tests/tensorbox/test_functions.py:
```diff
@@ -0,0 +1,519 @@
+            # watching makes all tensors trainable
+            tape.watch([t1, t2])
+            assert fn.requires_grad(t1)
+            assert fn.requires_grad(t2)
```
nice!

mariaschuld commented on tests/tensorbox/test_tensorbox.py:
```diff
@@ -0,0 +1,265 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Unit tests for TensorBox class"""
```
do i understand correctly, the tests of the base class are just done using numpy arrays?

mariaschuld replied:
hmm, wouldn't it be cleaner to test the numpy tensor box module instead? and just add a few minimal tests for the tensorbox base class?

josh146 replied:
I couldn't work out how, since it is an abstract base class and thus doesn't have all the methods implemented 🤔

mariaschuld commented on tests/tensorbox/test_tensorflow_box.py:
```diff
@@ -0,0 +1,214 @@
+
+    res = tape.gradient(y, x)
+    expected = tf.constant([[0.0, 0.0, 0.0], [8.0, 0.0, 0.0]])
+    assert np.allclose(res, expected)
```
nice test!

mariaschuld reviewed (approved):
Awesome work josh! 

Only major comment is to consider renaming the module to something revealing the purpose better.

Also, but i think this could be done subsequently, we should be clear about the overhead of wrapping!

josh146 commented on pennylane/__init__.py:
```diff
@@ -44,11 +44,11 @@
 from ._version import __version__
 from .io import *
 
+import pennylane.tensorbox  # pylint: disable=wrong-import-order
```
Yep :(

josh146 commented on pennylane/numpy/tensor.py:
```diff
@@ -132,6 +132,10 @@ def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
 
+        for i in args:
+            if hasattr(i, "unbox"):
+                return i.__array_ufunc__(ufunc, method, *args, **kwargs)
```
> Since the return statement will bypass the remainder of the for loop

This was somewhat deliberate. Here, I'm checking if any of the arrays are `TensorBox` objects implicitly (I cannot import the tensorbox module here as the NumPy module is much lower down the hierarchy, and it would cause circular imports). If an object in the input _is_ a tensor box, that we instead send this ufunc request to The `TensorBox.__array_ufunc__` method.

This is an issue with how NumPy dispatches a `ufunc`. If multiple arguments of the ufunc define custom `__array_ufunc__` methods, then we cannot know for sure which one will be called (it depends on the order of the input). See the discussion in https://numpy.org/devdocs/user/basics.subclassing.html#array-ufunc-for-ufuncs:

> E.g., lets assume that we evalulate `np.add(a, b)`, where b is an instance of another class B that has an override. If you use super as in the example, `ndarray.__array_ufunc__` will notice that b has an override, which means it cannot evaluate the result itself. Thus, it will return `NotImplemented` and so will our class A. Then, control will be passed over to b, which either knows how to deal with us and produces a result, or does not and returns `NotImplemented`, raising a `TypeError`.

co9olguy replied:
I think I've got it, thanks @josh146

josh146 commented on pennylane/tensorbox/fn.py:
```diff
@@ -0,0 +1,416 @@
+
+    non_numpy_interfaces = set(interfaces) - {"numpy"}
+
+    if len(non_numpy_interfaces) == 2:
```
this function is quite hardcoded 🤔 I think your comment also applies to line 52 (where we are checking if `tf` and `torch` are both in interfaces).

Perhaps better to treat `autograd` as 'special', and explicitly check the following two cases:

* No autograd, but more than one other interface - raise error

* Autograd and one other interface - warning to instead use vanilla numpy 

That is, change the logic as follows:

```python
if len(set(interfaces) - {"numpy", "autograd"}) > 1:
    # contains multiple non-autograd interfaces
    raise ValueError("Tensors contain mixed types; cannot determine dispatch library")

if len(set(interfaces) - {"numpy"}) > 1:
    # contains autograd and another interface
    warnings.warn(...)
```

co9olguy replied:
I just felt that checking against an integer rather than a set-like object seemed a weird way to do it

josh146 commented on pennylane/tensorbox/fn.py:
```diff
@@ -0,0 +1,416 @@
+    return np.all(t1 == t2, **kwargs)
+
+
+def allclose(a, b, rtol=1e-05, atol=1e-08, **kwargs):
```
yep

josh146 commented on pennylane/tensorbox/fn.py:
```diff
@@ -0,0 +1,416 @@
+    >>> cast(x, y)
+    tensor([1., 2.])
+    """
+    dtype = TensorBox(tensor2).numpy().dtype.type
```
I find numpy's dtype hierarchy quite confusing

josh146 commented on pennylane/tensorbox/fn.py:
```diff
@@ -0,0 +1,416 @@
+    >>> cast(x, y)
+    <tf.Tensor: shape=(2,), dtype=int64, numpy=array([1, 2])>
+    """
+    return TensorBox(tensor2).astensor(tensor1)
```
I don't think so. We're converting tensor1 to the same type as tensor2. 

That's why we wrap tensor2 in a TensorBox; we need to determine the correct dispatch library. E.g.:

1. `tensor1` is a numpy array, `tensor2` is a TF tensor
2. `tb = TensorBox(tensor2)` after inspecting the argument creates a `TensorFlowBox`
3. `tb.astensor(t)` is now a static method that converts its argument to a TensorFlow tensor

co9olguy replied:
Another argument in favour of the new UI :laughing:

josh146 commented on pennylane/tensorbox/fn.py:
```diff
@@ -0,0 +1,416 @@
+
+    Calling this function on a PennyLane NumPy array:
+
+    >>> x = np.array([1., 5.], requires_grad=True)
+    >>> qml.tensorbox.requires_grad(x)
+    True
+    >>> x.requires_grad = False
+    >>> qml.tensorbox.requires_grad(x)
+    False
```
```suggestion
    >>> x = np.array([1., 5.], requires_grad=True)
    >>> requires_grad(x)
    True
    >>> x.requires_grad = False
    >>> requires_grad(x)
    False
```

josh146 commented on pennylane/tensorbox/fn.py:
```diff
@@ -0,0 +1,416 @@
+    >>> y = tf.Variable([0.1, 0.2, 0.3])
+    >>> z = np.array([5., 8., 101.])
+    >>> stack([x, y, z])
+    TensorBox: <tf.Tensor: shape=(3, 3), dtype=float32, numpy=
```
Yep, sorry, this was a typo!
```suggestion
    <tf.Tensor: shape=(3, 3), dtype=float32, numpy=
```

josh146 commented on pennylane/tensorbox/fn.py:
```diff
@@ -0,0 +1,416 @@
+    array([[1, 3],
+           [2, 4]], dtype=int32)>
+    """
+    return TensorBox(tensor).T.data
```
Do you mean like?

```python
def T(tensor):
    mod = t.__class__.__module__.split(".")[0]
    if mod == "numpy":
        return t.T
    if mod == "tensorflow":
        import tensorflow as tf
        return tf.transpose(tensor)
    if mod == "torch":
        import torch
        return torch.transpose(tensor)
```

I suppose it comes down to whether a python class instantiation is significant overhead compare to a python function call, or whether the difference is negligible.

josh146 replied:
One major downside to the above though is that we lose `ufunc` compatibility. So we cannot create high-level functions like the above that include ufuncs, even something as small as addition/multiplication would need to branch per interface.

josh146 commented on pennylane/tensorbox/numpy_box.py:
```diff
@@ -29,37 +29,38 @@ def __init__(self, tensor):
 
-        return res
+    def expand_dims(self, axis):
+        return NumpyBox(np.expand_dims(self.data, axis=axis))
```
Yep, no copying of any memory takes place

josh146 commented on pennylane/tensorbox/tensorbox.py:
```diff
@@ -55,10 +48,9 @@ class TensorBox(abc.ABC):
+    TensorBox: <tf.Variable 'Variable:0' shape=(3,) dtype=float32, numpy=array([0.4, 0.1, 0.5], dtype=float32)>
 
-    The original tensor is available via the :meth:`~.unbox()` method:
+    The original tensor is available via the :meth:`~.unbox` method:
```
```suggestion
    The original tensor is available via the :meth:`~.unbox` method or the :attr:`data` attribute:
```

josh146 commented on pennylane/tensorbox/tensorbox.py:
```diff
@@ -70,13 +62,15 @@ class TensorBox(abc.ABC):
 
     >>> y.ones_like().expand_dims(0)
-    tf.Tensor([[1. 1. 1.]], shape=(1, 3), dtype=float32)
+    tf.Tensor([[1., 1., 1.]], shape=(1, 3), dtype=float32)
```
```suggestion
    tf.Tensor([[1. 1. 1.]], shape=(1, 3), dtype=float32)
```

josh146 commented on pennylane/tensorbox/tensorbox.py:
```diff
@@ -195,13 +265,20 @@ def expand_dims(self, axis):
+    @property
+    @abc.abstractmethod
+    def interface(self):
+        """str, None: The package that the :class:`.TensorBox` class
```
Yep! This will show up as an 'attribute' in the documentation.

Sphinx will interpret this exactly the same as if we had a `Returns:` block, but it avoids the redundant description of the returns argument.

I wonder if this also works for small methods?

josh146 commented on pennylane/tensorbox/tensorbox.py:
```diff
@@ -251,3 +346,8 @@ def stack(values, axis=0):
+    @property
+    @abc.abstractmethod
+    def T(self):
+        """Returns the transpose of the tensor."""
```
yep!

josh146 commented on pennylane/tensorbox/tf_box.py:
```diff
@@ -15,43 +15,61 @@
+
+    @staticmethod
+    def stack(values, axis=0):
+        res = tf.stack(TensorFlowBox.unbox_list(values), axis=axis)
+        return TensorFlowBox(res)
```
I originally did, and `black` made it look really ugly 🤔

josh146 commented on pennylane/tensorbox/torch_box.py:
```diff
@@ -13,39 +13,61 @@
+
+    @property
+    def shape(self):
+        return tuple(self.data.shape)
```
I can't remember; it either returned a list, or something even stranger, like a torch tensor lol.

josh146 commented on tests/tensorbox/test_functions.py:
```diff
@@ -0,0 +1,519 @@
+            res = res.numpy()
+            t = t.numpy()
+
+        assert onp.issubdtype(onp.asarray(t).dtype, onp.integer)
```
This tests that the original tensor has an integer dtype. The reason I don't compare to a _specific_ integer dtype (such as `np.int64`) is that the different interfaces have different rules as to which integer dtype they cast to 🙁 But since we are casting to a float, it is sufficient to simply check that the input was not a float (it's an integer)

josh146 commented on tests/tensorbox/test_functions.py:
```diff
@@ -0,0 +1,519 @@
+    def test_cast_torch_dtype(self):
+        """If the tensor is a Torch tensor, casting using a Torch dtype
+        will also work"""
+        t = torch.tensor([1, 2, 3])
```
```suggestion
        t = torch.tensor([1, 2, 3], dtype=torch.int64)
```

josh146 commented on tests/tensorbox/test_functions.py:
```diff
@@ -0,0 +1,519 @@
+        assert t2.shape == new_shape
+
+    def test_expand_dims_array(self, shape, axis, new_shape):
+        """Test that the expand dimensions works correctly
```
```suggestion
        """Test that expand_dimensions works correctly
```

josh146 commented on tests/tensorbox/test_tensorbox.py:
```diff
@@ -11,531 +11,255 @@
+    assert np.all(res.unbox() == x / y)
+
+    res = 5 / yT
+    assert np.all(res.unbox() == 5 / y)
```
added!

josh146 commented on tests/tensorbox/test_functions.py:
```diff
@@ -0,0 +1,519 @@
+        if hasattr(res, "numpy"):
+            res = res.numpy()
+
+        assert fn.allclose(res, np.array([[1, 3], [2, 4]]))
```
done!

co9olguy commented:
Still not crazy about the name `TensorBox` :smile:

josh146 commented:
Open to suggestions! I like Box because it is short and sweet, and describes the object. Other options I was thinking about were `TensorContainer`, but it became too long to keep typing out (e.g., `TensorFlowContainer`), and Black kept inserting annoying breaks.
